homeassistant/components/event/__init__.py

```diff
@@ -0,0 +1,185 @@
+    "ATTR_EVENT_TYPE",
+    "ATTR_EVENT_TYPES",
+    "DOMAIN",
+    "PLATFORM_SCHEMA_BASE",
+    "PLATFORM_SCHEMA",
+    "EventDeviceClass",
+    "EventEntity",
+    "EventEntityDescription",
+    "EventEntityFeature",
+]
+
+# mypy: disallow-any-generics
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Event entities."""
+    component = hass.data[DOMAIN] = EntityComponent[EventEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[EventEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[EventEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+@dataclass
+class EventEntityDescription(EntityDescription):
+    """A class that describes event entities."""
+
+    device_class: EventDeviceClass | None = None
+    event_types: list[str] | None = None
+
+
+class EventEntity(RestoreEntity):
+    """Representation of a Event entity."""
+
+    entity_description: EventEntityDescription
+    _attr_device_class: EventDeviceClass | None
+    _attr_event_types: list[str]
+    _attr_state: None
+    _attr_extra_state_attributes: None  # type: ignore[assignment]
+
+    __last_event: datetime | None = None
+    __last_event_type: str | None = None
+    __last_event_extra_state_attributes: dict[str, Any] | None = None
+
+    @property
+    def device_class(self) -> EventDeviceClass | None:
+        """Return the class of this entity."""
+        if hasattr(self, "_attr_device_class"):
+            return self._attr_device_class
+        if hasattr(self, "entity_description"):
+            return self.entity_description.device_class
+        return None
+
+    @property
+    def event_types(self) -> list[str]:
+        """Return a list of possible events."""
+        if hasattr(self, "_attr_event_types"):
+            return self._attr_event_types
+        if (
+            hasattr(self, "entity_description")
+            and self.entity_description.event_types is not None
+        ):
+            return self.entity_description.event_types
+        raise AttributeError()
+
+    @final
+    @callback
+    def async_trigger_event(
+        self, event_type: str, extra_state_attributes: dict[str, Any] | None = None
+    ) -> None:
+        """Process a new event."""
+        if event_type not in self.event_types:
+            raise ValueError(f"Invalid event type {event_type} for {self.entity_id}")
+        self.__last_event = dt_util.utcnow()
+        self.__last_event_type = event_type
+        self.__last_event_extra_state_attributes = extra_state_attributes
+        self.async_write_ha_state()
```

Comment:
I would not expect that the state is written here. There might be other attributes an integration want to set on the entity before writing the state.

May be the name of the coroutine could be better `async_register_event`?

Reply:
E.g. mqtt first processes all topic (they may be shared) and finally calls `self.async_write_ha_state()` to update the state.

Reply:
> E.g. mqtt first processes all topic (they may be shared) and finally calls self.async_write_ha_state() to update the state.

So, call this method last? 🤷

Reply:
> So, call this method last? 🤷

May be, but it is at least different from the other entities. May be add an option that you to choose if the state is written?

Reply:
There is nothing else adding anything to the state? So, I don't understand your comment I guess.

Reply:
Well an MQTT entity can be available or can be `unknown` or not and might have extra state attributes (that can be blocked btw). How is this for the `event` entity?

Reply:
The available proeprty can still be set on update if you'd like. The entity will go unavailable at that point. As that doesn't require a trigger to be called, the method this comment is on, isn't relevant.

The unknown state only exists once in an event entity, during its first creation. The unknown state isn't managed by the integration implementation.

Reply:
Removed it 👍